### DISCUSSION TITLE
Update external values file won't update HelmRelease

### DISCUSSION THREAD

**Discussion by Thaval** [fluxcd/flux2#2715] [Q&A]
Hi,

earlier, I posted a question on how to sync the values in the HelmRelease. I figured out I had to change the values inside the HelmRelease. 
However, we decided to maintain the values.yaml externally, so our HelmRelease looks like this:
```
kind: HelmRelease
apiVersion: helm.toolkit.fluxcd.io/v2beta1
metadata:
  namespace: flux-system
  name: myname
spec:
  upgrade:
    remediation:
      remediateLastFailure: true
      retries: 1
  interval: 1m0s
  rollback:
    timeout: 1m0s
    cleanupOnFail: true
    force: true
  install:
    remediation:
      retries: 1
  chart:
    spec:
      valuesFiles:
      - path/to//values.yaml
      version: '*'
      reconcileStrategy: Revision
      sourceRef:
        kind: GitRepository
        name: gitname
      chart: path/to/charts

```

I observed that changes made in "path/to/values.yaml" will not be reflected in the pod. The HelmRelease will be reconciled however, but the changes are not taken. Is this because everything is hosted in Git and not in a HelmRepository?
How could I solve this? We do not want to have the values inside the HelmRelease yaml.

**Comment by kingdonb** [comment]
When you set `reconcileStrategy: Revision` the changes in `valuesFiles` are expected to be reconciled with each new change.

The behavior you described usually is reported by folks who haven't found `reconcileStrategy`. It is strange that you don't see it. That should be enough to opt into the behavior where every change in `values.yaml` or any other template file is reflected immediately, rather than waiting for a new version number in `Chart.yaml`. The behavior that people don't usually understand is that `valuesFiles` really are considered as part of a chart, and each revision of a Chart is immutable, so you cannot get a new revision without a new version number. The `reconcileStrategy: revision` setting fixes this, ostensibly.

There are usually two good solutions for this (the one you mentioned, and...)

https://fluxcd.io/docs/guides/helmreleases/#refer-to-values-in-configmaps-generated-with-kustomize

This way you can externalize the configuration from the HelmChart and guarantee an upgrade happens when the configuration is changed. Please take a look and let us know if this solves your issue!

If it does not, please let us know what specific version of Flux is installed (please post the output from `flux check`)

**Reply by kingdonb** [reply]
In this example:

https://fluxcd.io/docs/guides/helmreleases/#refer-to-values-inside-the-chart

The files `values.yaml` and `values-prod.yaml` are both referenced with a relative path that starts from within the chart directory.

You mentioned also @Thaval that your `values.yaml` reference was of the form `path/to//values.yaml` – is it possible that you've pointed at a values file which does not exist, and that's why the changes are not picked up? (It seems unlikely, as you would have probably noticed that your settings from that value are not included in the first release revision. Just checking for obvious things before we look for bigger issues...)

**Reply by Thaval** [reply]
Thanks for your response. I'll have a closer look at your answer tomorrow. 
For now, I can say that the chart is hosted in the git repo and not in a HelmRepository. 
When the HelmRelease creates the chart, i can see with "kubectl get helmcharts --all-namespaces" that the values files were merged into the chart. 
That's kind of problematic because I want to use the chart with different values, one value file per environment. 

I'll be back tomorrow.

**Reply by Thaval** [reply]
I am back. So for now, the values inside the HelmRelease are not updated when I change the values in the values.yaml, which is passed as a list entry in "valuesFiles" within HelmRelease. 
So I'm trying out the "ConfigMap" solution right now. 
If this doesn't work, I will need to think about switching to HelmRepository for the chart again. However, Microsoft only accepts OCI (at least we lose a whole UI when we dont use OCI)

**Reply by Thaval** [reply]
Following the guide 

https://fluxcd.io/docs/guides/helmreleases/#refer-to-values-in-configmaps-generated-with-kustomize

I end up with ` ConfigMap/podinfo-values-tddfk9kkh7 namespace not specified, error: the server could not find the requested resource`

**Reply by kingdonb** [reply]
Please follow the guide once exactly before attempting modification. You may have to add `namespace` in one of the resources.

Also, OCI Helm chart support is coming, you can find the RFC and a link to the PR where it's being implemented here to follow the progress: https://github.com/fluxcd/flux2/pull/2597#issuecomment-1097726182

Anyway, this behavior should be the same between GitRepository and HelmRepository charts.

When you place values files into the chart directory, they are merged according to which ones are listed in the HelmRelease – you can have multiple values files in there, and only select some. Be aware that if `values.yaml` is not listed then it will be omitted, which may be a divergence from how Helm itself handles values files, but understanding that is because these `valuesFiles` are included as part of the chart, not as an externally merged addon. You have to list `values.yaml` too, if it is a default set of values that are expected to be merged (usually with the environment-specific ones merged in listed after.)

You can still have multiple HelmCharts generated from the same GitRepository source. They will get adorned with some metadata indicating that they are different in the version number. Each HelmRelease generates its own HelmChart when it is reconciled against the GItRepository, if it lists a different set of valuesFiles.

**Comment by Thaval** [comment]
I followed the guide there exactly. However, the configmap is not created and the changed value in values.yaml is not reflected (I added a entry to my ConfigMap).
I simply copy pasted the guide 

kustomizeconfig.yaml
```
nameReference:
- kind: ConfigMap
  version: v1
  fieldSpecs:
  - path: spec/valuesFrom/name
    kind: HelmRelease
```

namespace.yaml
```
apiVersion: v1
kind: Namespace
metadata:
  name: myNamespace
```

kustomization.yaml
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: flux-system
resources:
- namespace.yaml
- git-repository.yaml
- helm-release.yaml
configMapGenerator:
  - name: podinfo-values
    files:
      - values.yaml=values.yaml
configurations:
  - kustomizeconfig.yaml
```

helm-release.yaml
```
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  labels:
    kustomize.toolkit.fluxcd.io/namespace: flux-system
    kustomize.toolkit.fluxcd.io/name: myNamespace
  name: myNamespace
  namespace: flux-system
spec:
  upgrade:
    remediation:
      retries: 1
      remediateLastFailure: true
  interval: 1m0s
  rollback:
    force: true
    timeout: 1m0s
    cleanupOnFail: true
  install:
    remediation:
      retries: 1
  valuesFrom:
    - kind: ConfigMap
      name: podinfo-values
  chart:
    spec:
      reconcileStrategy: Revision
      sourceRef:
        kind: GitRepository
        name: myNamespace
      version: '*'
      chart: path/to/chart


```

the reconciliation however is successful.

**Reply by Thaval** [reply]
Got it!
Forgot the namespace in the helm-release.yaml..
Thanks! :)

**Comment by Thaval** [comment]
I have to reject my answer. So it does update the config map with the values, but it doesn't update configmaps or for example a deployment which uses the value of a config map. As test I added a new config entry to a configmap and pass that key + value as env to a deployment. The created pod does not get the new config entry, as the config map is not updated with the values provided by values.yaml, althought the configmap for the values.yaml is updated. 

Edit: Reconciliation did not work, the configmap and pods are not updated with the new values provided in values.yaml. 
Maybe I'm misusing this? How can I troubleshoot that?

**Reply by kingdonb** [reply]
I am not certain if your issue is resolved here, but it's getting hard to follow the different threads in this discussion.

If you are still struggling with this, maybe we can help you in a more synchronous way (join us on Slack!)

The fluxcd.io website should have all the invite links at the bottom of the page, we are in [#flux](https://cloud-native.slack.com/channels/flux) on the CNCF Slack. If you repost your question there, I'm sure someone will be able to help!

**Reply by Thaval** [reply]
Yea, that's true :D 
I asked on Slack. I hope someone is able to answer my question :)
